### PR TITLE
Use dynamic client for resource creation.

### LINKED ISSUE #233: Use dynamic client instead of rest client
# Proposal

Currently, we use the Kubernetes REST client for all resource creation. Instead, I think we should switch to using the [dynamic client](https://godoc.org/k8s.io/client-go/dynamic).

Our current REST client implementation assumes that the underlying resource is meant for Kubernetes. By abstracting this to the dynamic client, we get a few benefits:

1. We can replace a majority of our REST client code with the dynamic client, which already has good support for CRDs. See https://soggy.space/namespaced-crds-dynamic-client/ for a simple example.
2. We can take advantage of the [fake](https://godoc.org/k8s.io/client-go/dynamic/fake) client for use in tests.
2. This makes supporting other operations like #231 for CRDs easy, since the dynamic client works pretty seamlessly for any registered resource.
3. Providers can bring their own types by following the dynamic interface. I was able to get a nifty demo working where Tekton Triggers were creating GCB Builds, but hypothetically this can be done for any k8s or external resource, so long it satisfies the interface.

Demo/PoC: https://github.com/tektoncd/triggers/compare/master...wlynch:dynamic-client#diff-57fed74cee793226b40869dac1113d50R194-R199 (WIP)

### tl;dr 

Turn this:
```go
uri := createRequestURI(apiVersion, apiResource.Name, namespace, apiResource.Namespaced)
result := r.RESTClient.Post().
	RequestURI(uri).
	Body([]byte(rt)).
	SetHeader("Content-Type", "application/json").
	Do()
```

Into this:
```go
gvr := schema.GroupVersionResource{
	Group:    apiResource.Group,
	Version:  apiResource.Version,
	Resource: apiResource.Name,
}
_, err = r.DynamicClient.Resource(gvr).Namespace(namespace).Create(data, metav1.CreateOptions{})
```

### PR DESCRIPTION
**Pull request by wlynch**
# Changes

This removes the usage of the REST client in favor of a clientset
router of dynamic clients per type. This router allows for configuration
of clients on a per type basis, allowing for customization and support
for arbitrary resources (so long as the dynamic client interface is
followed).

Other minor changes:
* Changes findAPIResource to be a receiver on Sink.
* Allows resource UID func to be overridable to allow deterministic UID
  generation in tests.

Fixes #233 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [n/a] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/triggers/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```
Triggers use dynamic clients for resource creation instead of REST clients.
```

### CODE REVIEW

File: pkg/sink/sink_test.go [hunk 1]
```diff
@@ -32,64 +32,62 @@ import (
...
 	triggerLabel  = triggersv1.GroupName + triggersv1.TriggerLabelKey
 	eventIDLabel  = triggersv1.GroupName + triggersv1.EventIDLabelKey
+
+	triggerName = "trigger"
+	eventID     = "12345"
 )
 
+func init() {
+	// Override UID generator for consistent test results.
+	template.UID = func() string { return eventID }
```
Comment (dibyom) [inline pkg/sink/sink_test.go]: :100:

File: pkg/client/dynamic/clientset/clientset.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
+package clientset
+
+import (
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+)
+
+// Clientset maps GroupVersionResources to underlying dynamic clients. If the
```
[the remark this answers is not in the document]
Comment (wlynch) [reply pkg/client/dynamic/clientset/clientset.go]: Yup! The idea is we can map types -> clients.

File: pkg/client/dynamic/clientset/clientset.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+// Add adds a new mapping for the given resource.
+func (r *Clientset) Add(resource schema.GroupVersionResource, client dynamic.Interface) {
+	r.config[resource] = client
+}
+
+// Resource returns the dynamic Resource for the given GVR. If not configured,
+// an error resource is returned.
+func (r *Clientset) Resource(resource schema.GroupVersionResource) dynamic.NamespaceableResourceInterface {
```
[the remark this answers is not in the document]
Comment (wlynch) [reply pkg/client/dynamic/clientset/clientset.go]: This is to be compatible with the dynamic.Interface type (https://godoc.org/k8s.io/client-go/dynamic#Interface). No errors actually get bubbled up until we call something in the ResourceInterface

**Pull request merged**